transcom/mymove
# B-20671 Parse TPPS Paid Invoice Report INT

## [Agility ticket](https://www13.v1host.com/USTRANSCOM38/story.mvc/Summary?oidToken=Story%3A982861)

## Summary

One of the reports that we will receive and process from TPPS is the TPPS Paid Invoice Report. This report will contain information on if/when/amount that TPPS paid HomeSafe.

This work was to take the csv file we will receive, parse it, and store that information into the database. We do not have the connection intricacies finalized, that will be done in [B-20560](https://www13.v1host.com/USTRANSCOM38/assetdetail.v1?number=B-20560). For now, I have added a note to the Agility ticket asking that the `SFTPTPPSPaidInvoiceReportPickupDirectory` variable be defined as part of the work on that ticket.

I added a test .csv file, an exact copy of the example file we got from TPPS. Since we don't actually have processing these hooked up yet, everything is simulated via tests. That test `.csv` file is parsed and then stored in the test database.

See summary section [here](https://github.com/transcom/mymove/pull/13382) for more information on the various types of reports we send/receive/process from TPPS.

Is there anything you would like reviewers to give additional scrutiny?
Please look over the tests thoroughly and identify any issues you see, as the tests are the only way we are testing this for now.

### How to test

There are two tests here to pay attention to - my intent with the tests was for `pkg/services/invoice/process_tpps_paid_invoice_report_test.go` to use a true test `.csv` file to show that processing the file from start to end will work. Then with the `pkg/edi/tpps_paid_invoice_report/parser_test.go`, to prove on a more granular level that the parsing of the file works - using a string representation of the `.csv` file

1. Access Agility and download the attachment with the Title of [TPPS paid settlement .csv 7/31/24 (#1 example to use for parsing)](https://www13.v1host.com/USTRANSCOM38/Default.aspx?Page=Widgets%2fDetails%2fAttachmentDetails&AssetWindow=true&AssetContext=Attachment%3a992546)
2. That file is directly from the SFTP directory, placed there by US bank. I used that file as the basis for all the parsing and storage tests.
3. Open that .csv file and take a look. It is not comma delimited and has some interesting spacing going on.
4. Verify that the file you downloaded matches the `pkg/services/invoice/tpps_paid_invoice_report_testfile.csv` that I added.
5. Make your way over to `pkg/services/invoice/process_tpps_paid_invoice_report_test.go`. Verify that this test uses the `tpps_paid_invoice_report_testfile.csv` file and that the `successfully proccesses a valid TPPSPaidInvoiceReport and stores it in the database ` test passes.
6. Now head to `pkg/edi/tpps_paid_invoice_report/parser_test.go`. Verify that the sampleTPPSPaidInvoiceString is the string representation of the `.csv` file, and that the `successfully parse simple TPPS Paid Invoice string` test passes.
